[SPARK-47227][DOCS] Improve documentation for Spark Connect (#45335)
apache/spark: docs/spark-connect-overview.md

  <img src="img/spark-connect-communication.png" title="Spark Connect communication" alt="Spark Connect communication" />
</p>

> ## What is changing with Spark Connect

Reviewer: This heading makes it sound like Spark Connect used to work differently and is now changing. That is not the case, is it?

I think a better heading would be something along the lines of: "How Spark Connect differs from "traditional" Spark programs"

I couldn't find a better word to describe non-Connect programs than "traditional", but you get the gist.

One of the main design goals of Spark Connect is to enable a full separation and
isolation of the client from the server. As a consequence, there are some changes
that developers need to be aware of when using Spark Connect:

1. The client does not run in the same process as the Spark driver. This means that
   the client cannot directly access and interact with the driver JVM to manipulate
   the execution environment. In particular, in PySpark, the client does not use Py4J
   and thus the accessing the private fields holding the JVM implementation of `DataFrame`,
   `Column`, `SparkSession`, etc. is not possible (e.g. `df._jdf`).

> 2. By design, the Spark Connect protocol is designed around the concepts of Sparks logical
>    plans as the abstraction to be able to declarative describe the operations to be executed

Reviewer: Nit: Avoid repeating "design"; tweak some grammar.

```suggestion
2. By design, the Spark Connect protocol uses Spark logical
   plans as the abstraction to declaratively describe the operations to be executed
```
